### PR TITLE
(CL) fix: update fee accumulator upon swaps

### PR DESCRIPTION
**Pull request by mattverse**
## What is the purpose of the change

Alternative for https://github.com/osmosis-labs/osmosis/issues/3890

Currently, fee accumulators do not properly get updated during swaps when crossing ticks.
This PR updates fee accumulators upon swaps so that it updates the fee accumulators to properly update according to the netLiquidity of the tick that it's crossing. 

## Brief Changelog
Update Fee Accumulator upon swaps according to the netLiquidity of the crossing tick

## Testing and Verifying
- New field has been added for testing the fee accumulator values in the existing swap tests
- New tests has been added for `crossTick` method

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: osmoutils/accum/accum.go
Comment (p0mvn) [inline osmoutils/accum/accum.go:78]: We should be never subtracting from an accumulator. It is always increasing by-design

File: x/concentrated-liquidity/tick.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:67]: Hmm I'm not sure why we would do this.

What needs to happen here is the following:
- update the tick's fee accumulator by subtracting the current tick's fee growth outside from the accumulator value ref: #3893 and https://github.com/osmosis-labs/osmosis/blob/3c5ed4e26f18169958a20ecd844161279351f3e4/x/concentrated-liquidity/tick.go#L65-L73

**Comment by mattverse** [comment]
@p0mvn I think this PR is r4r while I fix the weird error in e2e testing

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:435]:
```suggestion
				s.Require().Equal(accum.GetValue(), sdk.NewDecCoins(defaultAccumCoins).MulDec(sdk.NewDec(2)))
```

**Comment by mattverse** [comment]
@p0mvn do you have any ideas here why checks might be failing?

**Comment by p0mvn** [comment]
@mattverse I think `osmoutils` must be re-tagged since you updated the `accum` package API

File: x/concentrated-liquidity/swaps_test.go
Comment (p0mvn) [reply x/concentrated-liquidity/swaps_test.go:143]: Interesting. I thought this was already covered by tests. Let me look into this

Comment (p0mvn) [reply x/concentrated-liquidity/swaps_test.go:143]: Just looked into this more. I think that if we are expecting these values to be `sdk.NewDecCoins(nil)`, we should add such an assertion to make sure they do not have some other value be erroneously set.

Let me know what you think

**Comment by mattverse** [comment]
As @p0mvn mentioned, seems that the e2e is failing due to not having updatedtags for the accums package which is a part of osmoutils that we manage separately.  (cref https://github.com/osmosis-labs/osmosis/issues/4004)

Marking this PR as blocked by https://github.com/osmosis-labs/osmosis/pull/3906 for now until we come across consensus upon tagging the packages

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:152]: This has been addressed

File: osmoutils/accum/accum.go
Comment (mattverse) [reply osmoutils/accum/accum.go]: resolved

File: x/concentrated-liquidity/swaps_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:649]: I'm a bit worried about such checks because it is easy to forget with setting the value in the test case, and then we would skip validation. Can we require these values to always be set?

File: x/concentrated-liquidity/swaps_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:296]: nit: we have a variable for these values called `cl.EmptyCoins`.  Let's replace all occurrences

File: x/concentrated-liquidity/swaps_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/swaps_test.go:294]: It would be great to have some short comment explanations for each test case and each position about why these are the values we are expecting.

I'm working on swap fees right now and finding such comments extremely valuable

**Review by p0mvn — COMMENTED** [review]
Overall LGTM. Nice work

Had some minor suggestions

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:649]: sure, the only trade off here seeems like having more nil fields in the test case fields itself, but I do agree with you overall. Would remove these nil checks

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:296]: nice! Changed!

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:649]: Added these for the first position ticks

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:305]: Added!

File: osmoutils/accum/accum.go
Comment (czarcas7ic) [inline osmoutils/accum/accum.go:69]: Is this comment a TODO?

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:254]: Sorry, having trouble reconciling this with the comment.

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:403]: Same here, not sure how the comment translates to the expected upper and lower tick fee growth

File: x/concentrated-liquidity/swaps_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:875]: This comment makes more sense than the previous ones but I am not seeing where 3-1 comes from.

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:422]: What is the difference between the manipulation/charge fee above and here?

File: osmoutils/accum/accum.go
Comment (mattverse) [reply osmoutils/accum/accum.go:69]: This was a TODO but now has been deleted

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:254]: Ah thanks for the comment! Updated the comment to make this more understandable!

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:403]: ditto for above ^ changed as well!

File: x/concentrated-liquidity/swaps_test.go
Comment (mattverse) [reply x/concentrated-liquidity/swaps_test.go:875]: ditto for above ^! Changed as well! Let me know if this makes more sense now!

File: x/concentrated-liquidity/tick_test.go
Comment (mattverse) [reply x/concentrated-liquidity/tick_test.go:396]: I think you were referring to this field `manipulateTickFeeGrowth` as I noticed that poolToGet was getting utilized while the `manipulateTickFeeGrowth` was not getting utilized at all! Updated!

Comment (mattverse) [reply x/concentrated-liquidity/tick_test.go:396]: PoolToGet is used to set positions and ticks for the pool(the second test case uses this and thus expects error)

File: x/concentrated-liquidity/tick_test.go
Comment (mattverse) [reply x/concentrated-liquidity/tick_test.go:422]: Deleted the manipulation / charge fee logic!

**Review by p0mvn — APPROVED** [review]
Amazing. LGTM once CI passes

**Pull request merged**
